Add remaining code to complete MVP

mcmire opened the pull request:
All of the commits up to now have been adding code to support an MVP version of this tool. This commit ties up the MVP by adding a first rule, which merely verifies that a project has a `src/` directory, the entrypoint function to the tool, and the executable which kicks off the whole thing. This commit also adds a package script, `run-tool`, which can be used in development to quickly smoke test new changes.

<!--
Thanks for your contribution! Take a moment to answer these questions so that reviewers have the information they need to properly understand your changes:

* What is the current state of things and why does it need to change?
* What is the solution your changes offer and how does it work?

Are there any issues or other links reviewers should consult to understand this pull request better? For instance:

* Fixes #12345
* See: #67890
-->

Fixes #37.

mcmire committed:
Add remaining code to complete MVP

All of the commits up to now have been adding code to support an MVP
version of this tool. This commit ties up the MVP by adding a first
rule, which merely verifies that a project has a `src/` directory, the
entrypoint function to the tool, and the executable which kicks off the
whole thing. This commit also adds a package script, `run-tool`, which
can be used in development to quickly smoke test new changes.

Gudahtt commented on src/cli.ts:
```diff
@@ -0,0 +1,22 @@
+  process.exitCode = 1;
+});
+
+// vi: ft=typescript
```
What does this line mean?

mcmire replied:
Oops, this is a Vim modeline to force it to syntax-highlight this file as a TypeScript file, as for some reason it gets confused. This is a bit of a hack that I forgot I left in, so I can figure out some other way to do this.

mcmire replied:
Removed here: [3b18d0e](https://github.com/MetaMask/module-lint/pull/33/commits/3b18d0e342fb4d952c972537ba4dbed4f332f3bc)

Gudahtt commented on src/main.ts:
```diff
@@ -0,0 +1,250 @@
+  outputLogger: OutputLogger;
+}) {
+  let yargsFailure: { message: string; error: Error } | null = null;
+  /* istanbul ignore next: At the moment, there is no real way that Yargs could fail */
```
If failure is not possible, why has this been written to anticipate it?

Also, it seems that strict is enabled (I think it's on by default), so any unrecognized flag/command would be treated as an error. So it should be possible to encounter this.

Gudahtt replied:
Ah, I see the line below on the `_` option. This seems to suggest that it is accepting unrecognized arguments, contrary to what the docs suggest. I mist be missing something here.

mcmire replied:
I realized I hadn't actually tested failure cases. I added one [here]([3c5ec57](https://github.com/MetaMask/module-lint/pull/33/commits/3c5ec5715e2fa52b8251015030bd9fa1c3d64d49)), and that meant I could remove one of the ignore lines. I kept this one, but moved it down. I'm not sure how Yargs generates errors, but I figured that if `onFail` takes an `error` argument we should do _something_ exceptional with it. (I could have used it further down in the `if (yargsFailure)` block, but for some reason TypeScript wants to type `yargsFailure` as `never`. I'm not quite sure why at the moment. I'll try playing around with this further.)

Gudahtt replied:
I had missed that strict mode is disabled by default, OK. Using strict options makes sense though, good idea

Gudahtt commented on src/main.ts:
```diff
@@ -0,0 +1,250 @@
+
+  const yargs = createYargs(hideBin(argv))
+    .usage(USAGE_TEXT)
+    .help(false)
```
Why disable `--help`? With this disabled, how is it possible for `options.help` below to be set?

mcmire replied:
In order to write tests at the functional level, I need to control what gets output to stderr. I could mock `process.stderr` in the tests, but I don't want to completely shut out all output, because sometimes we like to add logging to help with debugging, and I thought that would have been annoying if that didn't work out of the box. To fix this, I disable Yargs's default `help` behavior and handle it myself. This still happens to work because I'm not using strict mode — I can't, otherwise I wouldn't be able to accept arbitrary arguments — so Yargs just puts unrecognized options in `argv`, and I can access them there.

That said, 1) even if I can't use strict mode, I can use strict _options_ mode, and 2) I've just noticed that setting `help(false)` removes `--help` from the usage output, so I should put it back.

Gudahtt commented on src/main.ts:
```diff
@@ -0,0 +1,250 @@
+    .exitProcess(false)
+    .fail(onFail);
+
+  const options = await yargs.parse();
```
I'm having trouble understanding the yargs docs on this method. How does this compare with using the `argv` like this:

```
const { argv } = createYargs(...)
```

Also my first time seeing this import called anything other than `yargs`; not sure I understand the choice there either.

mcmire replied:
If you use `yargs` this way, the type of `argv` will be `{ ... } | Promise<{ ... }}>`. The reason for this is detailed in the [TypeScript docs](https://github.com/yargs/yargs/blob/main/docs/typescript.md), and the recommended solution is to await `argv`. So we could do:

``` typescript
const options = await yargs.argv;
```

That looks strange, so I figured I'd call `parse()` here, even if it does the same thing.

I also need the return value of the chain, because I need to call `.getHelp()` on it later. So I can't destructure it to pull `argv` out immediately, but I need to give that return value a name.

It seems that the Yargs documentation largely prefers requiring `yargs/yargs` directly, but in a few cases, e.g. the TypeScript docs they call the default import `yargs`, and then call the thing that `yargs` returns the `parser`. Then sometimes they call the `parser` `yargs`, e.g. for the [`command`](https://yargs.js.org/docs/#api-reference-commandcmd-desc-builder-handler) option. So it's a bit inconsistent, but I can settle on `parser` I suppose.

I've made these changes here. Hopefully this makes more sense: [1b5854d](https://github.com/MetaMask/module-lint/pull/33/commits/1b5854d392a65822afcf4f296dab7c648045ee73)

Gudahtt commented on src/main.ts:
```diff
@@ -0,0 +1,250 @@
+  const yargs = createYargs(hideBin(argv))
+    .usage(USAGE_TEXT)
+    .help(false)
+    .string('_')
```
What necessitated this setting? I see that the docs say that this results in "non-hyphenated arguments being interpreted as strings, regardless of whether they resemble numbers." But we don't have any arguments yet, strings or otherwise

mcmire replied:
We do have arguments. You can run this tool two ways: if you pass arguments, they're assumed to be the short names of MetaMask repos, and the tool will run the linter against just those repos; otherwise, if you don't pass anything, it will use a default set of repos.

mcmire replied:
In any case, I added this line to see if I could get `_` typed as `string[]` instead of `(string | number)[]`, but this merely changed the type to `string & (string | number)[]`, which is strange. In any case, we don't need to do this, since we're typecasting `_` to strings below anyway, so I'll drop this.

mcmire replied:
Removed here: [dfba17e](https://github.com/MetaMask/module-lint/pull/33/commits/dfba17e7b5f4f2ae2cb6fbd653e8cca10b1391d1)

Gudahtt commented on src/constants.ts:
```diff
@@ -1,5 +1,127 @@
+export const USAGE_TEXT = `
+Analyzes one or more repos for divergence from a template repo.
+
+${THIS_PROJECT_NAME} OPTIONS [ARGUMENTS...]
```
Should this be

```suggestion
${THIS_PROJECT_NAME}
```

Since there are no options or arguments? Or maybe

```suggestion
${THIS_PROJECT_NAME} [--help]
```

mcmire replied:
There aren't any options, but there are arguments. I've clarified what kind of arguments they should be and removed the options: [1aed3b2](https://github.com/MetaMask/module-lint/pull/33/commits/1aed3b22c1ea9e594b5603c1d3d0d880ab3dd4c5)

Gudahtt commented on src/constants.ts:
```diff
@@ -1,5 +1,127 @@
+/**
+ * Wherever the tool was run.
+ */
+export const WORKING_DIRECTORY_PATH = process.cwd();
```
Not something I would describe as a "constant" :thinking: And only used in one place. Maybe better off inlining this.

mcmire replied:
Fair enough, inlined here: [732fc82](https://github.com/MetaMask/module-lint/pull/33/commits/732fc82c3f2d5e5cbe9147de6cd168e5ed098137)

Gudahtt commented on src/constants.ts:
```diff
@@ -1,5 +1,127 @@
+ */
+export const DEFAULT_CACHED_REPOSITORIES_DIRECTORY_PATH = path.join(
+  THIS_PROJECT_DIRECTORY_PATH,
+  'tmp/repositories',
```
For platform cross-compatibility, we shouldn't embed a delimiter here. Something like this should be equivalent except with better platform support:

```suggestion
  'tmp', 'repositories',
```

davidmurdoch replied:
Not sure how I ended up looking at this PR, but `path.join` normalizes paths to the path separator for the OS.

mcmire replied:
Ah, that's good. Well, it doesn't hurt, so I'll accept this suggestion.

Gudahtt commented on src/constants.ts:
```diff
@@ -1,5 +1,127 @@
+ * All of the remote MetaMask repositories that will be linted if a list is not
+ * explicitly provided.
+ */
+export const DEFAULT_PROJECT_NAMES = [
```
Curious how this list was generated

mcmire replied:
I generated this list from the code ownership spreadsheet. However, now that we have a new spreadsheet, perhaps I should use that instead? I wasn't sure what this should cover at the moment.

Gudahtt replied:
Hmm. I guess we can figure it out later.

I was wondering because for a lot of these, we explicitly don't want to make them match the template because we're not really maintaining them (e.g. for the `ethjs-` packages, it's not really a goal to incorporate them fully into our set of packages, that would be a ton of work). And in some cases, we intentionally diverge from the template because the package isn't a TypeScript library (e.g. the actions).

Gudahtt replied:
Maybe later on we could create an action template, and maintain a separate list for those

mcmire replied:
List scaled back here: [b33c692](https://github.com/MetaMask/module-lint/pull/33/commits/b33c69208da0f243c3d945b625f81e27ecf10291)

mcmire committed:
Drop .string('_'), as it doesn't add anything

mcmire committed:
Inline working directory path

mcmire committed:
Remove Vim modeline

mcmire committed:
Add --help back to usage output; use strict options mode

mcmire committed:
Use more widely used names for working with yargs

mcmire committed:
Remove OPTIONS from usage

mcmire committed:
Remove --version as an option

Because we've set `.exitProcess(false)`, this option no longer exits
automatically, so we'd have to handle it ourselves. I'm not sure anyone
will miss this (yet).

mcmire committed:
Remove (most of) the istanbul ignores

mcmire committed:
Build path in a guaranteed cross-platform way

Co-authored-by: Mark Stacey <markjstacey@gmail.com>

Gudahtt commented on src/main.ts:
```diff
@@ -0,0 +1,258 @@
+  const workingDirectoryPath = process.cwd();
+
+  const inputs = await parseInputs({ argv, outputLogger, defaultProjectNames });
+  /* istanbul ignore next: At the moment, there is no real way that Yargs could fail */
```
Is this branch still unreachable? Both here and in `parseInputs`, the branch where `null` is returned still has an ignore statement, but it looks like `parseCommandLineArguments` can return `null` (so it should be possible to test it here as well)

mcmire replied:
Ah, you're right, this is now reachable. Not sure why I thought it wasn't before now 🤔  Removed here: [af59f53](https://github.com/MetaMask/module-lint/pull/33/commits/af59f5384afb4a21477964d071b770406c94842e)

Gudahtt replied:
Thanks! I think you just missed it after turning on `strictOptions`

Gudahtt commented on src/main.ts:
```diff
@@ -0,0 +1,258 @@
+}) {
+  let yargsFailure: { message: string; error: Error } | null = null;
+  const onFail = (message: string, error: Error) => {
+    /* istanbul ignore next: Not sure how to produce this */
```
I think we'd need to actually _throw an error_ in a command to reproduce this. Assuming that's correct, it makes sense to leave this uncovered since it should be unreachable

mcmire replied:
Ah... okay. Makes sense!

Gudahtt commented on src/main.ts:
```diff
@@ -0,0 +1,258 @@
+    ...options,
+    // This is the name that Yargs gives to the arguments that aren't options.
+    // eslint-disable-next-line @typescript-eslint/naming-convention
+    _: options._.map(String),
```
Nit: Perhaps this is an opportunity to give this a better name?

mcmire replied:
Good point! Updated here: [ca73b38](https://github.com/MetaMask/module-lint/pull/33/commits/ca73b387df05edbead39a25364876906b10e83d4)

Gudahtt commented on src/main.ts:
```diff
@@ -0,0 +1,258 @@
+    (rejectedProjectLintResultPromiseOutcome) => {
+      outputLogger.logToStderr(
+        'stack' in rejectedProjectLintResultPromiseOutcome.reason
+          ? rejectedProjectLintResultPromiseOutcome.reason.stack
```
This seems.... a little strange to me. Why are we logging just the stack here, rather than the full error? `console.error` at least seems to format errors reasonable well (including the stack), and it automatically supports error causes too which is a bonus

mcmire replied:
That's fair! I've fixed the OutputLogger to handle error objects (well, non-strings in general) better: [d034718](https://github.com/MetaMask/module-lint/pull/33/commits/d034718c7f27b8a6ed1e24e48c6b89a855edbd29)

Gudahtt commented on src/main.ts:
```diff
@@ -0,0 +1,258 @@
+ * repositories will be (or have been) cloned.
+ * @param args.outputLogger - Writable streams for output messages.
+ */
+async function lintProjects({
```
Great idea to use `allSettled` here to handle failures gracefully, and to put failures last. Seems good.

Thoughts on adding a little summary at the end? e.g X/Y passed

mcmire replied:
Good idea, added here: [8a025a0](https://github.com/MetaMask/module-lint/pull/33/commits/8a025a06b5978bea7f13a295f6c54337af5719bf)

Gudahtt replied:
Nice!

Gudahtt commented on src/rules/helpers.ts:
```diff
@@ -0,0 +1,54 @@
+
+/**
+ * Rule objects are fairly abstract: the name of a rule and the dependencies of
+ * a rule can be anything; and unfortunately, we cannot really enforce this or
```
Sorry I'm having trouble following this sentence. It says that rules can be anything, but then that we can't enforce this. Enforce what? I thought rules could be anything.

mcmire replied:
Er, you're right. Does this make more sense? [5989401](https://github.com/MetaMask/module-lint/pull/33/commits/59894010c62aab5f640da15a1b0f6402e9f61761)

Gudahtt replied:
Not exactly; what does "enforce names" mean?

Gudahtt replied:
Oh, I guess you mean that it ensures that other referenced rules exist. Ok, I think that makes sense.

Gudahtt commented on src/rules/require-source-directory.ts:
```diff
@@ -0,0 +1,19 @@
+    const stats = await project.fs.getEntryStats(entryPath);
+    const passed = stats?.isDirectory();
+    const message = stats
+      ? `\`${entryPath}\` is not a directory when it should be.`
```
Nit: At this point we don't know that it's a file, that's tracked by the `passed` variable. I see that this is only used when `passed` is `false`, but maybe we could refactor this to make these more obviously connected?

mcmire replied:
Updated here. Is this clearer? [bc3f58e](https://github.com/MetaMask/module-lint/pull/33/commits/bc3f58e12c001be24380104f9e83fd78f38d1780)

Gudahtt commented on src/rules/require-source-directory.ts:
```diff
@@ -0,0 +1,19 @@
+    const passed = stats?.isDirectory();
+    const message = stats
+      ? `\`${entryPath}\` is not a directory when it should be.`
+      : `\`${entryPath}\` exists in the module template, but not in this repo.`;
```
Nit: Maybe we could simplify this message a bit

```suggestion
      : `\`${entryPath}\` does not exist in this repo.`;
```

The rule doesn't actually check the module template after all, so it might be confusing if we ever restructure the template

mcmire replied:
Updated here: [bc3f58e](https://github.com/MetaMask/module-lint/pull/33/commits/bc3f58e12c001be24380104f9e83fd78f38d1780)

Gudahtt commented on src/main.test.ts:
```diff
@@ -0,0 +1,437 @@
+            ],
+            validRepositories: [
+              {
+                name: 'repo-1',
```
Nit: Perhaps we should reverse the order here, and make repo-2 fail? Otherwise this test might pass even if the tool exited when encountering the failure (since the failure is on the last repo in the list)

mcmire replied:
Okay, changed here: [dda0390](https://github.com/MetaMask/module-lint/pull/33/commits/dda0390bad4dda628c0e5c8a9a317c24c27b0d9f)

mcmire committed:
Remove another 'istanbul ignore' directive

mcmire committed:
No need to hold on to the '_' name

mcmire committed:
Remove another 'istanbul ignore' directive

mcmire committed:
Fix formatting

mcmire committed:
Don't just log the error stack, log the whole thing

mcmire committed:
TEMP: Don't lint tmp/

mcmire committed:
Add passing/failing numbers to report

mcmire committed:
Clarify JSDocs for buildRule

mcmire committed:
Refactor require-source-directory and simplify message

mcmire committed:
Remove chalk

mcmire committed:
Remove extraneous line from main.test.ts

mcmire committed:
Fix tests

mcmire committed:
Add missing tests for require-source-directory

mcmire commented on src/main.test.ts:
```diff
@@ -0,0 +1,437 @@
+describe('main', () => {
+  beforeEach(() => {
+    fakeDateOnly();
+    execaMock.mockImplementation((): ExecaChildProcess => {
```
Ah, the `mockExeca` helper already handles providing a default implementation, so this line is extraneous. Removed here: [0ba1e29](https://github.com/MetaMask/module-lint/pull/33/commits/0ba1e290dad5cc1dda76b736d3ff008aa313b4b2)

mcmire committed:
Make the 'does not exit immediately' tests better

mcmire committed:
Limit the number of default repos

Gudahtt commented on src/rules/require-source-directory.ts:
```diff
@@ -0,0 +1,24 @@
+
+    const stats = await project.fs.getEntryStats(entryPath);
+
+    if (stats?.isDirectory()) {
+      return pass();
+    }
+
+    const message =
+      stats && !stats.isDirectory()
+        ? `\`${entryPath}/\` is not a directory when it should be.`
+        : `\`${entryPath}/\` does not exist in this project.`;
+    const failures = [{ message }];
+    return fail(failures);
```
Nit: We could eliminate the extra check by restructing it like this:

```suggestion
    if (!stats) {
      return fail([{ message: `\`${entryPath}/\` does not exist in this project.` }]);
    } else if (!stats.isDirectory()) {
      return fail([{ message: `\`${entryPath}/\` is not a directory when it should be.` }]);
    }

    return pass();
```

Or maybe this if we want to avoid repeating the `fail` call:

```suggestion
    const failures = [];
    if (!stats) {
      failures.push({ message: `\`${entryPath}/\` does not exist in this project.` });
    } else if (!stats.isDirectory()) {
      failures.push({ message: `\`${entryPath}/\` is not a directory when it should be.` }});
    }
    
    if (failures.length > 0) {
      return fail(failures);
    }
    return pass();
```

Gudahtt commented on src/execute-rules.ts:
```diff
@@ -224,7 +224,7 @@ async function executeRule({
  * in automatically).
  */
-function pass(): SuccessfulPartialRuleExecutionResult {
+export function pass(): SuccessfulPartialRuleExecutionResult {
```
Nit: Looks like these aren't tested directly yet

mcmire replied:
Good call — tested here: [475b273](https://github.com/MetaMask/module-lint/pull/33/commits/475b27343b2b3f67ee4e2542da3113d066a76e2a)

Gudahtt commented on src/output-logger.ts:
```diff
@@ -86,11 +86,14 @@ export class OutputLogger implements AbstractOutputLogger {
-  if (rest.length > 0) {
-    stream.write(format(`${messageOrFormatString}\n`, ...rest));
+export function logToStream(stream: SimpleWriteStream, args: unknown[]) {
+  if (typeof args[0] === 'string' && /%[sdifjoOc%]/u.test(args[0])) {
```
What does this regular expression do?

mcmire replied:
This checks for the existence of a format specifier, which `util.format` takes. I've extracted this check and documented it to make it more clear: [8b53d27](https://github.com/MetaMask/module-lint/pull/33/commits/8b53d273733f9ff0a69d21145de141addd2be984)

mcmire committed:
Refactor better

Co-authored-by: Mark Stacey <markjstacey@gmail.com>

mcmire committed:
Silence jest-haste-map errors

mcmire committed:
Add tests for pass and fail

mcmire committed:
Refactor the check in  and document it

mcmire committed:
Fix lint violation

Gudahtt reviewed (approved):
LGTM!
